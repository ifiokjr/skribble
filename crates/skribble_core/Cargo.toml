[package]
name = "skribble_core"
version = { workspace = true }
authors = { workspace = true }
categories = { workspace = true }
documentation = "https://docs.rs/skribble_core"
edition = { workspace = true }
homepage = { workspace = true }
include = ["src/**/*.rs", "Cargo.toml", "readme.md"]
keywords = ["build", "atomic", "css"]
license = { workspace = true }
readme = "readme.md"
repository = { workspace = true }
rust-version = { workspace = true }
description = "Core features for the skribble atomic css library"

[dependencies]
derivative = { workspace = true }
derive_more = { workspace = true }
doc-comment = { workspace = true }
globset = { workspace = true }
heck = { workspace = true }
indent_write = { workspace = true }
indexmap = { features = ["serde"], workspace = true }
lazy_static = { workspace = true }
lightningcss = { workspace = true }
readonly = { workspace = true }
regex = { workspace = true }
serde = { workspace = true }
serde_json = { workspace = true }
skribble_color = { workspace = true, features = ["serde"] }
thiserror = { workspace = true }
typed-builder = { workspace = true }
vfs = { workspace = true }

[dev-dependencies]
insta = { features = ["json"], workspace = true }
