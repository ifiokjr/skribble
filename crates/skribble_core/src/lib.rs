#![deny(clippy::all)]

pub use error::*;

mod error;
