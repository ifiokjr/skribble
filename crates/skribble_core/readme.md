# skribble_core

> core library for skribble

<br />

[![Crate][crate-image]][crate-link] [![Docs][docs-image]][docs-link] [![Status][ci-status-image]][ci-status-link] [![Unlicense][unlicense-image]][unlicense-link]

## Installation

```toml
[dependencies]
skribble_core = "0.0.0"
```

### Usage

```rust
use skribble_core::*;
```

[crate-image]: https://img.shields.io/crates/v/skribble_core.svg
[crate-link]: https://crates.io/crates/skribble_core
[docs-image]: https://docs.rs/skribble_core/badge.svg
[docs-link]: https://docs.rs/skribble_core
[ci-status-image]: https://github.com/ifiokjr/skribble/workflows/ci/badge.svg
[ci-status-link]: https://github.com/ifiokjr/skribble/actions?query=workflow:ci
[unlicense-image]: https://img.shields.io/badge/license-Unlicence-blue.svg
[unlicense-link]: https://opensource.org/license/unlicense
