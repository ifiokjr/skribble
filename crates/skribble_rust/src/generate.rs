use std::fmt::Write;

use indexmap::indexmap;
use indexmap::IndexMap;
use skribble_core::format_css_string;
use skribble_core::indent_writer;
use skribble_core::wrap_indent;
use skribble_core::AnyEmptyResult;
use skribble_core::AnyResult;
use skribble_core::Atom;
use skribble_core::LinkedValues;
use skribble_core::Prioritized;
use skribble_core::ToSkribbleCss;

use super::indoc;
use super::RunnerConfig;
use super::ToPascalCase;
use super::ToSnakeCase;
use crate::error::Error;

fn generate_media_queries(
  config: &RunnerConfig,
  method_names: &mut IndexMap<String, String>,
  sections: &mut Vec<String>,
  struct_names_map: &mut IndexMap<String, usize>,
  trait_names: &mut Vec<String>,
) -> AnyEmptyResult {
  for (key, map) in config.media_queries.iter() {
    let mut section = Vec::<String>::new();
    let trait_name = format!("GeneratedMediaQuery{}", key.to_pascal_case());
    let struct_name = format!("{trait_name}Child");
    section.push(generate_struct(&struct_name));
    section.push(generate_impl_skribble_value(&struct_name));

    let mut methods = vec![format!("pub trait {trait_name}: GeneratedSkribbleValue {{")];

    for (name, media_query) in map.iter() {
      let method_name = get_method_name(name, GLOBAL_PREFIX, method_names)?;
      let css_docs = wrap_indent(
        wrap_docs(wrap_in_code_block(
          media_query_docs(&media_query.query),
          "css",
        )),
        1,
      );

      if let Some(ref description) = media_query.description {
        methods.push(wrap_indent(wrap_docs(description), 1));
        methods.push(wrap_indent(wrap_docs("\n"), 1));
      }

      methods.push(css_docs);
      methods.push(wrap_indent(
        format!("#[inline]\nfn {method_name}(&self) -> {struct_name} {{"),
        1,
      ));
      methods.push(wrap_indent(
        format!("{struct_name}::from_ref(self.append(\"{name}\"))"),
        2,
      ));
      methods.push(wrap_indent("}", 1));
      methods.push(wrap_indent(
        format!(
          "#[inline]\nfn {method_name}_(&self, property: &'static str, value: &'static str) -> \
           String {{"
        ),
        1,
      ));
      methods.push(wrap_indent(
        format!("self.append(format!(\"{name}:[{{}}={{}}]\", property.trim(), value.trim()))"),
        2,
      ));
      methods.push(wrap_indent("}", 1));
    }

    methods.push("}".into());
    section.push(methods.join("\n"));

    trait_names.push(trait_name);
    struct_names_map.insert(struct_name, trait_names.len());
    sections.push(section.join("\n"));
  }

  Ok(())
}

fn media_query_docs(query: impl AsRef<str>) -> String {
  let query = query.as_ref();
  format!("@media {query} {{\n  /* ... */\n}}")
}

fn modifier_docs(values: &[String]) -> String {
  let value = values.join(", ");
  format!("{value} {{\n  /* ... */\n}}")
}

fn generate_modifiers(
  config: &RunnerConfig,
  method_names: &mut IndexMap<String, String>,
  sections: &mut Vec<String>,
  struct_names_map: &mut IndexMap<String, usize>,
  trait_names: &mut Vec<String>,
) -> AnyEmptyResult {
  for (key, map) in config.modifiers.iter() {
    let mut section = Vec::<String>::new();
    let trait_name = format!("GeneratedModifier{}", key.to_pascal_case());
    let struct_name = format!("{trait_name}Child");
    section.push(generate_struct(&struct_name));
    section.push(generate_impl_skribble_value(&struct_name));

    let mut methods = vec![format!("pub trait {trait_name}: GeneratedSkribbleValue {{")];

    for (name, modifier) in map.iter() {
      let method_name = get_method_name(name, GLOBAL_PREFIX, method_names)?;
      let css_docs = wrap_indent(
        wrap_docs(wrap_in_code_block(modifier_docs(&modifier.values), "css")),
        1,
      );

      if let Some(ref description) = modifier.description {
        methods.push(wrap_indent(wrap_docs(description), 1));
        methods.push(wrap_indent(wrap_docs("\n"), 1));
      }

      methods.push(css_docs);
      methods.push(wrap_indent(
        format!("#[inline]\nfn {method_name}(&self) -> {struct_name} {{"),
        1,
      ));
      methods.push(wrap_indent(
        format!("{struct_name}::from_ref(self.append(\"{name}\"))"),
        2,
      ));
      methods.push(wrap_indent("}", 1));

      methods.push(wrap_indent(
        format!(
          "#[inline]\nfn {method_name}_(&self, property: &'static str, value: &'static str) -> \
           String {{"
        ),
        1,
      ));
      methods.push(wrap_indent(
        format!("self.append(format!(\"{name}:[{{}}={{}}]\", property.trim(), value.trim()))"),
        2,
      ));
      methods.push(wrap_indent("}", 1));
    }

    methods.push("}".into());
    section.push(methods.join("\n"));

    trait_names.push(trait_name);
    struct_names_map.insert(struct_name, trait_names.len());
    sections.push(section.join("\n"));
  }

  Ok(())
}

fn generate_keyframes(
  config: &RunnerConfig,
  method_names: &mut IndexMap<String, String>,
  sections: &mut Vec<String>,
) -> AnyEmptyResult {
  sections.push(format!(
    "pub trait {KEYFRAME_TRAIT_NAME}: GeneratedSkribbleValue {{"
  ));

  for (name, keyframe) in config.keyframes.iter() {
    let method_name = get_method_name(name, KEYFRAMES_PREFIX, method_names)?;
    let css_docs = wrap_indent(
      wrap_docs(wrap_in_code_block(keyframe.to_skribble_css(config)?, "css")),
      1,
    );

    if let Some(ref description) = keyframe.description {
      sections.push(wrap_indent(wrap_docs(description), 1));
      sections.push(wrap_indent(wrap_docs("\n"), 1));
    }

    sections.push(css_docs);

    sections.push(wrap_indent(
      format!("#[inline]\nfn {method_name}(&self) -> String {{"),
      1,
    ));

    sections.push(wrap_indent(format!("self.append_value(\"{name}\")"), 2));

    sections.push(wrap_indent("}", 1));
  }

  sections.push("}".into());

  Ok(())
}

fn generate_named_classes(
  config: &RunnerConfig,
  method_names: &mut IndexMap<String, String>,
  sections: &mut Vec<String>,
  trait_names: &mut Vec<String>,
) -> AnyEmptyResult {
  sections.push("pub trait GeneratedNamedClasses: GeneratedSkribbleValue {".into());

  for (class_name, class) in config.classes.iter() {
    let method_name = get_method_name(class_name, GLOBAL_PREFIX, method_names)?;

    if let Some(ref description) = class.description {
      sections.push(wrap_indent(wrap_docs(description), 1));
    }

    sections.push(wrap_indent(
      format!("#[inline]\nfn {method_name}(&self) -> String {{"),
      1,
    ));

    sections.push(wrap_indent(
      format!("self.append_value(\"{class_name}\")"),
      2,
    ));

    sections.push(wrap_indent("}", 1));
  }

  trait_names.push("GeneratedNamedClasses".into());
  sections.push("}".into());

  Ok(())
}

fn generate_atoms(
  config: &RunnerConfig,
  method_names: &mut IndexMap<String, String>,
  sections: &mut Vec<String>,
  trait_names: &mut Vec<String>,
) -> AnyEmptyResult {
  let mut struct_content = Vec::<String>::new();
  let mut trait_content = Vec::<String>::new();

  // parent modifiers
  trait_content.push(format!(
    "pub trait {ATOM_TRAIT_NAME}: GeneratedSkribbleValue {{"
  ));

  for (name, atom) in config.atoms.iter() {
    let method_name = get_method_name(name, GLOBAL_PREFIX, method_names)?;
    let atom_struct_name = format!("GeneratedAtom{}", name.to_pascal_case());

    struct_content.push(generate_struct(&atom_struct_name));
    struct_content.push(generate_impl_skribble_value(&atom_struct_name));

    match atom.values {
      LinkedValues::Color => {
        struct_content.push(format!(
          "impl {COLOR_TRAIT_NAME} for {atom_struct_name} {{}}"
        ));
      }
      LinkedValues::Keyframes => {
        struct_content.push(format!(
          "impl {KEYFRAME_TRAIT_NAME} for {atom_struct_name} {{}}",
        ));
      }
      LinkedValues::Values(ref value_sets) => {
        let value_set_trait_name = get_value_set_trait_name(name);
        generate_atom_value_sets(
          config,
          atom,
          value_sets,
          &value_set_trait_name,
          &mut struct_content,
          method_names,
        )?;

        struct_content.push(format!(
          "impl {value_set_trait_name} for {atom_struct_name} {{}}",
        ));
      }
    }

    // The atom
    if let Some(ref description) = atom.description {
      trait_content.push(wrap_indent(wrap_docs(description), 1));
    }

    trait_content.push(wrap_indent(
      format!("#[inline]\nfn {method_name}(&self) -> {atom_struct_name} {{"),
      1,
    ));

    trait_content.push(wrap_indent(
      format!("{atom_struct_name}::from_ref(self.append(\"{name}\"))"),
      2,
    ));

    trait_content.push(wrap_indent("}", 1));

    // The atom argument
    trait_content.push(wrap_indent(
      format!("#[inline]\nfn {method_name}_(&self, value: &'static str) -> String {{"),
      1,
    ));

    trait_content.push(wrap_indent(
      format!("self.append(format!(\"{name}:[{{}}]\", value.trim()))"),
      2,
    ));

    trait_content.push(wrap_indent("}", 1));
  }

  trait_content.push("}".into());

  trait_names.push(ATOM_TRAIT_NAME.into());
  sections.push(struct_content.join("\n"));
  sections.push(trait_content.join("\n"));

  Ok(())
}

fn generate_atom_value_sets(
  config: &RunnerConfig,
  atom: &Atom,
  value_sets: &skribble_core::NameSet,
  value_set_trait_name: impl AsRef<str>,
  struct_content: &mut Vec<String>,
  method_names: &mut IndexMap<String, String>,
) -> AnyEmptyResult {
  let value_set_trait_name = value_set_trait_name.as_ref();
  let atom_name = &atom.name;

  struct_content.push(format!(
    "pub trait {value_set_trait_name}: GeneratedSkribbleValue {{"
  ));

  for Prioritized { value, .. } in value_sets.iter() {
    let Some(value_set) = config.value_sets.get(value) else {
        continue;
      };

    for (value_name, _) in value_set.values.iter() {
      let method_name = get_method_name(value_name, atom_name, method_names)?;
      let mut css = String::new();
      let mut css_properties = indent_writer();
      let css_name = format_css_string(atom_name);
      let css_suffix = if value_name.is_empty() {
        "".into()
      } else {
        format!(r"\:\${}", format_css_string(value_name))
      };
      writeln!(css, ".{css_name}{css_suffix} {{")?;
      atom.write_css_properties(&mut css_properties, config, value_name)?;
      write!(css, "{}", css_properties.get_ref())?;
      writeln!(css, "}}")?;

      let css_docs = wrap_indent(wrap_docs(wrap_in_code_block(css, "css")), 1);
      struct_content.push(css_docs);

      struct_content.push(wrap_indent(
        format!("#[inline]\nfn {method_name}(&self) -> String {{"),
        1,
      ));

      struct_content.push(wrap_indent(
        format!("self.append_value(\"{value_name}\")"),
        2,
      ));

      struct_content.push(wrap_indent("}", 1));
    }
  }
  struct_content.push("}".into());
  Ok(())
}

fn generate_atom_colors(
  config: &RunnerConfig,
  method_names: &mut IndexMap<String, String>,
  sections: &mut Vec<String>,
) -> AnyEmptyResult {
  sections.push(format!(
    "pub trait {COLOR_TRAIT_NAME}: GeneratedSkribbleValue {{"
  ));

  for (name, css_variable) in config.css_variables.iter() {
    if !css_variable.is_color() {
      continue;
    }

    let method_name = get_method_name(name, COLORS_PREFIX, method_names)?;
    let mut property_rule = String::new();
    css_variable.write_property_rule(&mut property_rule, config, false)?;
    let css_docs = wrap_indent(wrap_docs(wrap_in_code_block(property_rule, "css")), 1);

    if let Some(ref description) = css_variable.description {
      sections.push(wrap_indent(wrap_docs(description), 1));
      sections.push(wrap_indent(wrap_docs("\n"), 1));
    }

    sections.push(css_docs);

    sections.push(wrap_indent(
      format!("#[inline]\nfn {method_name}(&self) -> String {{"),
      1,
    ));
    sections.push(wrap_indent(format!("self.append_value(\"{name}\")"), 2));
    sections.push(wrap_indent("}", 1))
  }

  for (name, _) in config.palette.iter() {
    let method_name = get_method_name(name, COLORS_PREFIX, method_names)?;

    sections.push(wrap_indent(
      format!("#[inline]\nfn {method_name}(&self) -> String {{"),
      1,
    ));

    sections.push(wrap_indent(format!("self.append_value(\"{name}\")"), 2));

    sections.push(wrap_indent("}", 1));
  }

  sections.push("}".into());

  Ok(())
}

fn generate_css_variables(
  config: &RunnerConfig,
  method_names: &mut IndexMap<String, String>,
  sections: &mut Vec<String>,
) -> AnyEmptyResult {
  let mut entries = vec![
    indoc!(
      "
    pub fn vars() -> GeneratedCssVariables {
      GeneratedCssVariables
    }
    pub struct GeneratedCssVariables;
    impl GeneratedCssVariables {"
    )
    .into(),
  ];

  for (name, css_variable) in config.css_variables.iter() {
    let method_name = get_method_name(name, VARIABLES_PREFIX, method_names)?;
    let variable_name = css_variable.get_variable(config.options());
    let mut property_rule = String::new();
    css_variable.write_property_rule(&mut property_rule, config, false)?;
    let css_docs = wrap_indent(wrap_docs(wrap_in_code_block(property_rule, "css")), 1);

    if let Some(ref description) = css_variable.description {
      entries.push(wrap_indent(wrap_docs(description), 1));
      entries.push(wrap_indent(wrap_docs("\n"), 1));
    }

    entries.push(css_docs.clone());

    entries.push(wrap_indent(
      format!("#[inline]\npub fn {method_name}(&self) -> String {{"),
      1,
    ));

    entries.push(wrap_indent(format!("\"{variable_name}\".into()",), 2));

    entries.push(wrap_indent("}", 1));
  }

  entries.push("}".into());
  sections.push(entries.join("\n"));

  Ok(())
}

fn wrap_docs(content: impl AsRef<str>) -> String {
  let mut result = vec![];
  for line in content.as_ref().lines() {
    result.push(format!("/// {line}"));
  }

  result.join("\n")
}

fn wrap_in_code_block(content: impl AsRef<str>, r#type: impl AsRef<str>) -> String {
  format!(
    "```{}\n{}\n```",
    r#type.as_ref(),
    content.as_ref().trim_end(),
  )
}

fn get_value_set_trait_name(value_set_name: impl Into<String>) -> String {
  format!(
    "GeneratedValueSet{}",
    value_set_name.into().to_pascal_case()
  )
}

fn generate_impl_skribble_value(name: impl AsRef<str>) -> String {
  format!(
    indoc!(
      "
      impl GeneratedSkribbleValue for {} {{
        #[inline]
        fn from_ref(value: impl AsRef<str>) -> Self {{
          Self(value.as_ref().to_string())
        }}
        #[inline]
        fn get_skribble_value(&self) -> &String {{
          &self.0
        }}
      }}"
    ),
    name.as_ref()
  )
}

fn generate_struct_implementations(
  struct_names_map: &IndexMap<String, usize>,
  trait_names: &[String],
  sections: &mut Vec<String>,
) {
  let mut content = Vec::<String>::new();
  for (struct_name, min_index) in struct_names_map.iter() {
    for (index, trait_name) in trait_names.iter().enumerate() {
      if *min_index <= index {
        content.push(format!("impl {trait_name} for {struct_name} {{}}",));
      }
    }
  }

  sections.push(content.join("\n"));
}

fn generate_struct(name: impl AsRef<str>) -> String {
  let name = name.as_ref();
  format!("pub struct {name}(String);")
}

fn get_method_name(
  name: impl AsRef<str>,
  prefix: impl AsRef<str>,
  method_names: &mut IndexMap<String, String>,
) -> AnyResult<String> {
  let method_name = safe_method_name(&name)?;
  let mut index = 0;
  let mut current_method_name = method_name.clone();

  loop {
    let with_prefix = format!("{}:::{}", prefix.as_ref(), current_method_name);
    if method_names.contains_key(&with_prefix) {
      index += 1;
      current_method_name = format!("{}_{}", method_name, index);
      continue;
    }

    method_names.insert(with_prefix, name.as_ref().to_string());
    break;
  }

  Ok(current_method_name)
}

fn safe_method_name(name: impl AsRef<str>) -> Result<String, Error> {
  let name = name.as_ref();

  let prefix = match name.chars().next() {
    Some(first_char) if first_char.is_ascii_digit() => "n",
    Some(first_char) if !first_char.is_ascii_alphabetic() => {
      match first_char {
        '_' => "u",
        '-' => "m",
        '+' => "p",
        '.' => "d",
        _ => return Err(Error::InvalidMethodName(name.to_string())),
      }
    }
    _ => "",
  };

  let method_name = format!("{prefix}{}", name.to_snake_case());

  if RESERVED_WORDS.contains(&method_name.as_str()) {
    return Ok(format!("r#{}", method_name));
  }

  if method_name.is_empty() {
    return Ok("__".into());
  }

  Ok(method_name)
}

const RESERVED_WORDS: &[&str] = &[
  "abstract", "alignof", "as", "become", "box", "break", "const", "continue", "crate", "do",
  "else", "enum", "extern", "false", "final", "fn", "for", "if", "impl", "in", "let", "loop",
  "macro", "match", "mod", "move", "mut", "offsetof", "override", "priv", "proc", "pub", "pure",
  "ref", "return", "Self", "self", "sizeof", "static", "struct", "super", "trait", "true", "type",
  "typeof", "unsafe", "unsized", "use", "virtual", "where", "while", "yield",
];
const ATOM_TRAIT_NAME: &str = "GeneratedAtom";
const KEYFRAME_TRAIT_NAME: &str = "GeneratedKeyframeSet";
const COLOR_TRAIT_NAME: &str = "GeneratedColorSet";
pub(crate) const GLOBAL_PREFIX: &str = "global";
pub(crate) const COLORS_PREFIX: &str = "colors";
pub(crate) const KEYFRAMES_PREFIX: &str = "keyframes";
pub(crate) const VARIABLES_PREFIX: &str = "variables";

const HEADER: &str = r#"#![allow(unused)]
// This file was generated by skribble.
use private::GeneratedSkribbleValue;
pub fn sk() -> GeneratedSkribbleRoot {
  GeneratedSkribbleRoot::from_ref("")
}
pub struct GeneratedSkribbleRoot(String);
impl GeneratedSkribbleValue for GeneratedSkribbleRoot {
  #[inline]
  fn from_ref(value: impl AsRef<str>) -> Self {
    Self(value.as_ref().to_string())
  }
  #[inline]
  fn get_skribble_value(&self) -> &String {
    &self.0
  }
}
mod private {
  #[doc(hidden)]
  pub trait GeneratedSkribbleValue {
    fn from_ref(value: impl AsRef<str>) -> Self;
    fn get_skribble_value(&self) -> &String;
    #[inline]
    fn append(&self, value: impl AsRef<str>) -> String {
      let current_value = self.get_skribble_value();
      let prefix = if current_value.is_empty() {
        "".into()
      } else {
        format!("{current_value}:")
      };

      format!("{}{}", prefix, value.as_ref())
    }
    #[inline]
    fn append_value(&self, value: impl AsRef<str>) -> String {
      format!("{}:${}", self.get_skribble_value(), value.as_ref())
    }
  }
}"#;

fn combine_sections_with_header(sections: Vec<String>) -> String {
  format!("{HEADER}\n{}", sections.join("\n"))
}

pub(crate) fn generate_file_contents(
  config: &RunnerConfig,
) -> AnyResult<(String, IndexMap<String, String>)> {
  let mut method_names = IndexMap::<String, String>::new();

  let mut sections = Vec::<String>::new();
  let mut trait_names = vec![];
  let mut struct_names_map: IndexMap<String, usize> =
    indexmap! { "GeneratedSkribbleRoot".into() => 0 };

  generate_css_variables(config, &mut method_names, &mut sections)?;
  generate_media_queries(
    config,
    &mut method_names,
    &mut sections,
    &mut struct_names_map,
    &mut trait_names,
  )?;
  generate_modifiers(
    config,
    &mut method_names,
    &mut sections,
    &mut struct_names_map,
    &mut trait_names,
  )?;
  generate_keyframes(config, &mut method_names, &mut sections)?;
  generate_atom_colors(config, &mut method_names, &mut sections)?;
  generate_atoms(config, &mut method_names, &mut sections, &mut trait_names)?;
  generate_named_classes(config, &mut method_names, &mut sections, &mut trait_names)?;
  generate_struct_implementations(&struct_names_map, &trait_names, &mut sections);

  Ok((combine_sections_with_header(sections), method_names))
}
