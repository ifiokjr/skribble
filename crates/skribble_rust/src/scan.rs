use std::fmt::Write;

use indexmap::IndexMap;
use rstml::node::Node;
use rstml::node::NodeAttribute;
use rstml::node::NodeBlock;
use rstml::parse2;
use skribble_core::AnyResult;
use skribble_core::Arguments;
use skribble_core::AtomType;
use skribble_core::ClassFactory;
use skribble_core::Classes;
use skribble_core::RunnerConfig;
use syn::visit;
use syn::visit::Visit;
use syn::File;

use crate::generate::COLORS_PREFIX;
use crate::generate::GLOBAL_PREFIX;
use crate::generate::KEYFRAMES_PREFIX;
use crate::generate::TRANSFORMER_PREFIX;

#[readonly::make]
struct ScanVisitor<'config, 'names> {
  pub classes: Classes,
  config: &'config RunnerConfig,
  method_names: &'names IndexMap<String, String>,
}

impl<'config, 'names> ScanVisitor<'config, 'names> {
  pub fn new(
    config: &'config RunnerConfig,
    method_names: &'names IndexMap<String, String>,
  ) -> Self {
    Self {
      classes: Classes::default(),
      config,
      method_names,
    }
  }

  fn update_with_tokens(&mut self, tokens: &[String], transformers: &[String], arguments: &String) {
    if !starts_with_sk(tokens) {
      return;
    }

    let mut factory = ClassFactory::new(self.config);
    let arguments = if arguments.is_empty() {
      None
    } else {
      Some(Arguments::from(arguments))
    };

    if let Some(tokens) = tokens.get(1..) {
      for token in tokens {
        let token_identifier =
          if token == "__" && arguments.as_ref().map(|args| args.is_kv()).unwrap_or(false) {
            continue;
          } else if token.ends_with('_') {
            token.trim_end_matches('_')
          } else {
            token
          };

        let key = match factory.get_atom_type() {
          Some(AtomType::Color) => {
            format!("{COLORS_PREFIX}:::{token_identifier}")
          }
          Some(AtomType::Keyframes) => {
            format!("{KEYFRAMES_PREFIX}:::{token_identifier}")
          }
          Some(_) => {
            if let Some(atom) = factory.get_atom() {
              format!("{atom}:::{token_identifier}")
            } else {
              String::new()
            }
          }
          None => format!("{GLOBAL_PREFIX}:::{token_identifier}"),
        };

        let token = if let Some(value) = self.method_names.get(&key) {
          value
        } else {
          token_identifier
        };

        factory.add_token(token);
      }
    }

    for transformer in transformers.iter() {
      factory.add_transformer(&transformer.into());
    }

    if let Some(value) = arguments {
      factory.add_argument(value);
    }

    self.classes.insert_factory(factory);
  }

  fn visit_leptos_view_macro(&mut self, node: &syn::Macro) -> bool {
    // check if this is a leptos view (by name for now)
    let Some(view_segment) = node.path.segments.last() else {
      return false;
    };

    if view_segment.ident != "view" {
      return false;
    }

    let Ok(nodes) = parse2(node.tokens.clone()) else {
      return false;
    };

    self.visit_leptos_nodes(&nodes);
    true
  }

  fn visit_leptos_nodes(&mut self, nodes: &[Node]) {
    for node in nodes.iter() {
      match node {
        Node::Fragment(fragment) => self.visit_leptos_nodes(&fragment.children),
        Node::Element(element) => {
          self.visit_leptos_nodes(&element.children);

          let attributes = element.attributes();

          for attribute in attributes.iter() {
            match attribute {
              NodeAttribute::Block(NodeBlock::ValidBlock(block)) => self.visit_block(block),
              NodeAttribute::Attribute(attribute) => {
                let Some(keyed_attribute) = attribute.possible_value.as_ref() else {
                  continue;
                };

                self.visit_expr(&keyed_attribute.value);
              }
              _ => {}
            }
          }
        }
        Node::Block(NodeBlock::ValidBlock(block)) => {
          self.visit_block(block);
        }

        _ => {}
      }
    }
  }

  fn update_skribble_method(&mut self, node: &syn::ExprMethodCall) -> bool {
    let mut tokens = vec![node.method.to_string()];
    let mut arguments = String::new();
    let mut transformers = vec![];
    read_arguments_from_method_call(node, &mut arguments);
    read_tokens_from_expression(
      self.method_names,
      node.receiver.as_ref(),
      &mut tokens,
      &mut transformers,
    );
    self.update_with_tokens(&tokens, &transformers, &arguments);

    starts_with_sk(&tokens)
  }
}

impl<'ast, 'config, 'names> Visit<'ast> for ScanVisitor<'config, 'names> {
  fn visit_expr_method_call(&mut self, node: &'ast syn::ExprMethodCall) {
    if !self.update_skribble_method(node) {
      visit::visit_expr_method_call(self, node);
    }
  }

  fn visit_macro(&mut self, node: &'ast syn::Macro) {
    if self.visit_leptos_view_macro(node) {
      return;
    }

    visit::visit_macro(self, node);
  }
}

fn read_arguments_from_method_call(method: &syn::ExprMethodCall, arguments: &mut dyn Write) {
  match (method.args.len(), method.args.first(), method.args.last()) {
    (1, Some(syn::Expr::Lit(lit)), _) => {
      if let syn::Lit::Str(value_literal) = &lit.lit {
        write!(arguments, "{}", value_literal.value()).unwrap();
      }
    }
    (2, Some(syn::Expr::Lit(property)), Some(syn::Expr::Lit(value))) => {
      if let (syn::Lit::Str(property_literal), syn::Lit::Str(value_literal)) =
        (&property.lit, &value.lit)
      {
        write!(
          arguments,
          "{}={}",
          property_literal.value(),
          value_literal.value()
        )
        .ok();
      }
    }
    _ => {}
  }
}

fn read_tokens_from_expression(
  method_names: &IndexMap<String, String>,
  node: &syn::Expr,
  tokens: &mut Vec<String>,
  transformers: &mut Vec<String>,
) {
  match node {
    syn::Expr::MethodCall(method) => {
      let method_name = method.method.to_string();
      let mut arguments = String::new();
      read_arguments_from_method_call(method, &mut arguments);
      let key = format!("{}:::{method_name}", TRANSFORMER_PREFIX);

      if let Some(transformer) = method_names.get(&key) {
        if arguments.is_empty() {
          transformers.push(transformer.into());
        } else {
          transformers.push(format!("{transformer}={arguments}"));
        }
      } else {
        tokens.insert(0, method_name);
      }

      read_tokens_from_expression(method_names, method.receiver.as_ref(), tokens, transformers);
    }
    syn::Expr::Call(call) => {
      read_tokens_from_expression(method_names, call.func.as_ref(), tokens, transformers);
    }
    syn::Expr::Path(path) => {
      if let Some(path) = path
        .path
        .segments
        .last()
        .map(|segment| segment.ident.to_string())
      {
        tokens.insert(0, path);
      }
    }
    _ => {}
  }
}

pub(crate) fn scan(
  config: &RunnerConfig,
  _file_path: impl AsRef<str>,
  content: impl AsRef<str>,
  method_names: &IndexMap<String, String>,
) -> AnyResult<Classes> {
  let mut visitor = ScanVisitor::new(config, method_names);
  let syntax_tree: File = syn::parse_str(content.as_ref())?;

  visitor.visit_file(&syntax_tree);

  Ok(visitor.classes)
}

fn starts_with_sk(tokens: &[String]) -> bool {
  tokens.first() == Some(&"sk".to_string())
}
