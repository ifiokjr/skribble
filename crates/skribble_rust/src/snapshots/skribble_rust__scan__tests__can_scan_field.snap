---
source: crates/skribble_rust/src/scan.rs
expression: classes.to_skribble_css(config)?
---
/* Generated by Skribble */
@charset "utf-8";
@layer base, alias, default;
@layer default {
  .p\:\$px {
    padding: 1px;
  }
}

